src/objectc.c

```diff
@@ -222,6 +222,200 @@ void msgpack_object_print(FILE* out, msgpack_object o)
+        aux_buffer = aux_buffer + ret;
+        aux_buffer_size = aux_buffer_size - ret;
+        break;
+
+    case MSGPACK_OBJECT_BOOLEAN:
+        ret = snprintf(aux_buffer, aux_buffer_size, (o.via.boolean ? "true" : "false"));
+        aux_buffer = aux_buffer + ret;
+        aux_buffer_size = aux_buffer_size - ret;
+        break;
+
+    case MSGPACK_OBJECT_POSITIVE_INTEGER:
+#if defined(PRIu64)
+        ret = snprintf(aux_buffer, aux_buffer_size, "%" PRIu64, o.via.u64);
+        aux_buffer = aux_buffer + ret;
+        aux_buffer_size = aux_buffer_size - ret;
+#else
+        if (o.via.u64 > ULONG_MAX) {
+            ret = snprintf(aux_buffer, aux_buffer_size, "over 4294967295");
+            aux_buffer = aux_buffer + ret;
+            aux_buffer_size = aux_buffer_size - ret;
+        } else {
+            ret = snprintf(aux_buffer, aux_buffer_size, "%lu", (unsigned long)o.via.u64);
+            aux_buffer = aux_buffer + ret;
+            aux_buffer_size = aux_buffer_size - ret;
+        }
+#endif
+        break;
+
+    case MSGPACK_OBJECT_NEGATIVE_INTEGER:
+#if defined(PRIi64)
+        ret = snprintf(aux_buffer, aux_buffer_size, "%" PRIi64, o.via.i64);
+        aux_buffer = aux_buffer + ret;
+        aux_buffer_size = aux_buffer_size - ret;
+#else
+        if (o.via.i64 > LONG_MAX) {
+            ret = snprintf(aux_buffer, aux_buffer_size, "over +2147483647");
+            aux_buffer = aux_buffer + ret;
+            aux_buffer_size = aux_buffer_size - ret;
+        } else if (o.via.i64 < LONG_MIN) {
+            ret = snprintf(aux_buffer, aux_buffer_size, "under -2147483648");
+            aux_buffer = aux_buffer + ret;
+            aux_buffer_size = aux_buffer_size - ret;
+        } else {
+            ret = snprintf(aux_buffer, aux_buffer_size, "%ld", (signed long)o.via.i64);
+            aux_buffer = aux_buffer + ret;
+            aux_buffer_size = aux_buffer_size - ret;
+        }
+#endif
+        break;
+
+    case MSGPACK_OBJECT_FLOAT:
+        ret = snprintf(aux_buffer, aux_buffer_size, "%f", o.via.f64);
+        aux_buffer = aux_buffer + ret;
+        aux_buffer_size = aux_buffer_size - ret;
+        break;
+
+    case MSGPACK_OBJECT_STR:
+        ret = snprintf(aux_buffer, aux_buffer_size, "\"");
+        aux_buffer = aux_buffer + ret;
+        aux_buffer_size = aux_buffer_size - ret;
+        ret = snprintf(aux_buffer, aux_buffer_size, "%.*s", o.via.str.size, o.via.str.ptr);
+        aux_buffer = aux_buffer + ret;
+        aux_buffer_size = aux_buffer_size - ret;
+        ret = snprintf(aux_buffer, aux_buffer_size, "\"");
+        aux_buffer = aux_buffer + ret;
+        aux_buffer_size = aux_buffer_size - ret;
+        break;
+
+    case MSGPACK_OBJECT_BIN:
+        ret = snprintf(aux_buffer, aux_buffer_size, "\"");
+        aux_buffer = aux_buffer + ret;
+        aux_buffer_size = aux_buffer_size - ret;
+        ret = snprintf(aux_buffer, aux_buffer_size, "%.*s", o.via.bin.size, o.via.bin.ptr);
```

Comment:
See also the [C standard draft](http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf), §7.19.6.1, in the description of the %s conversion specifier: "If the precision is specified, no more than that many bytes are written. If the precision is not specified or is greater than the size of the array, the array shall contain a null character".

This also means your string printing works because although it has no null character, the precision is exactly the array size.